### ISSUE TITLE
Stop gradient for some entry of a tensor

### ISSUE THREAD

**Issue by JerrikEph** [tensorflow/tensorflow#9162]
As far as I know, the tf.stop_gradient function can only stop the gradient of a whole tensor. I recently wanted to implement a model that requires the stop of gradient for some entry of a tensor (not the whole tensor). and I have come up with an workaround,
```
def entry_stop_gradients(target, mask):
    mask_h = tf.logical_not(mask)
    
    mask = tf.cast(mask, dtype=target.dtype)
    mask_h = tf.cast(mask_h, dtype=target.dtype)
    
    return tf.stop_gradient(mask_h * target) + mask * target
```
 
hope somebody could implement a low level version of this feature.

**Comment by asimshankar** [comment]
This question is better asked on  [StackOverflow](http://stackoverflow.com/questions/tagged/tensorflow) since it is not a  bug or feature request. There is also a larger community that reads questions there. Thanks!

**Comment by JerrikEph** [comment]
It is in fact a feature request. Tf already have a stop_gradient function that stop the gradient for a tensor as a whole, why not implement a generalized version?

**Comment by ashwhall** [comment]
I'm on board with this. It would be nice if there were an overloaded version of tf.stop_grads like the example above.

**Comment by ashwhall** [comment]
Is this going to remain closed? It's definitely a feature request, not something that could be aided by StackOverflow.

**Comment by jasonyoun** [comment]
I wonder why this remains closed. I think this is a necessary feature to be added.

**Comment by asimshankar** [comment]
Does [`tf.custom_gradient`](https://github.com/tensorflow/tensorflow/blob/r1.7/tensorflow/python/ops/custom_gradient.py#L33) (in 1.7, `tf.contrib.eager.custom_gradient` in 1.6) help?

You can use that to define the custom gradient that appropriately masks out portions that are of interest.

If not, the workaround suggested in the original posting seems reasonable. 
I'm perhaps misunderstanding what you'd want as a feature.

**Comment by 7GrandPa** [comment]
Hi Jerrik,  if the mask is sparse(# of 0s is much greater than # of nonzeros), will your approach accelerate the backprop speed, because lots of gradients are no calculate?

**Comment by JerrikEph** [comment]
@7GrandPa 
No, it won't. since it actually introduces another gradient masked identity function. more computation is required actually.

**Comment by Sanqiang** [comment]
I think asimshankar's solution: custom_gradient works

**Comment by alexcoca** [comment]
@asimshankar Can you give an example of how `tf.custom_gradient` would achieve it? As far as I know, you can't use the autodiff to differentiate wrt to a `tf.Variable` slice. So in the backwards pass, all you can do is to just multiply the incoming gradients by the mask? Is that what you are suggesting?

**Comment by a-berg** [comment]
For tensorflow 2.X, wouln't just computing the gradient then multiplying it by the mask before applying it be enough? I don't know if this would have much overhead, but e.g.:
```python
loss = lambda: (z_obj - model(X))**2  # squared error, I am not interested in the "mean" for my use case.
with tf.GradientTape() as tape:
        z = loss()
    grads = tape.gradient(z, X)
    optim.apply_gradients(zip([grads*mask], [X]))
```
I checked it on a small optimization problem and works as intended (optimizes in the available directions, leaving the rest unchanged).
